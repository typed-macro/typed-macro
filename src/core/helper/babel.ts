import template from '@babel/template'
import traverse from '@babel/traverse'
import * as types from '@babel/types'
import { parse, parseExpression } from '@babel/parser'
import generate from '@babel/generator'

export type Babel = {
  /**
   * From @babel/template.
   * Help generate nodes from code string.
   *
   * e.g.
   * ```typescript
   * template.statement.ast('const a = 1') // generate a statement node
   * ```
   */
  template: typeof template

  /**
   * From @babel/traverse.
   * Help traverse from a node.
   *
   * e.g.
   * ```typescript
   * traverse(someNode, {
   *   CallExpression(path) {
   *     ...
   *   }
   * })
   * ```
   */
  traverse: typeof traverse

  /**
   * From @babel/types.
   * Help check node type and create nodes of specific type.
   *
   * e.g.
   * ```typescript
   *  types.isIdentifier(someNode) // check is identifier
   *  types.identifier('hello') // create identifier node
   * ```
   */
  types: typeof types

  /**
   * From @babel/parser.
   * Help parse code string to ast.
   *
   * e.g.
   * ```typescript
   *  parse('const a = 1')
   * ```
   */
  parse: typeof parse

  /**
   * From @babel/parser.
   * Help parse code string to expression.
   *
   * e.g.
   * ```typescript
   *  parseExpression('1 + 2')
   * ```
   */
  parseExpression: typeof parseExpression

  /**
   * From @babel/generator.
   * Help generate code string from ast.
   *
   * e.g.
   * ```typescript
   *  generate(someNode)
   * ```
   */
  generate: typeof generate
}

export const BABEL_TOOLS: Babel = Object.freeze({
  template,
  traverse,
  types,
  parse,
  parseExpression,
  generate,
})
