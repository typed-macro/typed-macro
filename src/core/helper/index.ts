export { BABEL_TOOLS, Babel } from './babel'
export { ImportOption } from './import'
export { getPathHelper, PathHelper } from './path'
export { getStateHelper, State, StateHelper } from './state'
export { getTransformHelper, TransformHelper } from './transform'
