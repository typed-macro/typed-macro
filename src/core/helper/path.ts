import { dirname, resolve, sep } from 'path'
import { existsSync } from 'fs'

export type PathPatternNormalizer = {
  /**
   * Normalized path or path pattern,
   */
  normalized: string
  /**
   * Base path for the normalized path or path pattern.
   */
  base: string
  /**
   * Get import path from file path got by normalized path and base,
   * almost the opposite operation of normalization
   * @param path file path
   */
  resolveImportPath: (path: string) => string
}

export function normalizePathPattern(
  pattern: string,
  importer: string,
  root: string
): PathPatternNormalizer {
  if (!pattern.startsWith('.') && !pattern.startsWith('/')) {
    throw new Error(
      `pattern must start with '.' or '/' (relative to project root)`
    )
  }
  let base: string
  let parentDepth = 0
  const isAbsolute = pattern.startsWith('/')
  if (isAbsolute) {
    base = resolve(root)
    pattern = pattern.slice(1)
  } else {
    base = dirname(importer)
    while (pattern.startsWith('../')) {
      pattern = pattern.slice(3)
      base = resolve(base, '../')
      parentDepth++
    }
    if (pattern.startsWith('./')) {
      pattern = pattern.slice(2)
    }
  }
  return {
    normalized: pattern,
    base,
    resolveImportPath: isAbsolute
      ? (f) => `/${f}`
      : parentDepth
      ? (f) => `${'../'.repeat(parentDepth)}${f}`
      : (f) => `./${f}`,
  }
}

const projectDirCache = {
  root: '',
  leaf: '',
}

export function projectDir(which: 'root' | 'leaf') {
  if (projectDirCache[which]) return projectDirCache[which]

  const paths = (require.main?.path || process.cwd()).split(sep)

  if (which === 'root') {
    for (let i = 0; i < paths.length; i++) {
      const path = paths.slice(0, i + 1).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['root'] = path)
    }
  } else {
    for (let i = paths.length; i > 0; i--) {
      const path = paths.slice(0, i).join(sep)
      if (existsSync([path, 'node_modules'].join(sep)))
        return (projectDirCache['leaf'] = path)
    }
  }

  throw new Error('can not find project root')
}
