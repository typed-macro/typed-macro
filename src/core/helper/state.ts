export type State = {
  /**
   * Get the state shared during traversal.
   */
  get<T, K>(key: K): T

  /**
   * Set the state shared during traversal.
   */
  set<T, K>(key: K, value: T): T

  /**
   * Clear the state shared during traversal.
   */
  clear(): boolean

  /**
   * Delete the state shared during traversal.
   */
  delete<K>(key: K): void
}

export function createState(): State {
  const state = new Map<any, any>()
  return {
    get(key) {
      return state.get(key)
    },
    set(key, value) {
      state.set(key, value)
      return value
    },
    clear() {
      state.clear()
      return true
    },
    delete<K>(key: K) {
      state.delete(key)
    },
  }
}
