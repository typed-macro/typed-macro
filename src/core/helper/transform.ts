import { Node, NodePath } from '@babel/traverse'
import template from '@babel/template'
import { File, ImportDeclaration, Program } from '@babel/types'
import { ImportOption, matchImportStmt, renderImportStmt } from './import'
import { findProgramPath } from '@/common'

export type TransformHelper = {
  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement,
   *   or undefined if imports param is array and length === 0
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   *   or undefined if imports param is array and length === 0
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Get current program or the program of provided node.
   * @param node any node path
   */
  getProgram: (node?: NodePath) => NodePath<Program>
}

export function getTransformHelper(ast: File): TransformHelper {
  const thisProgram = findProgramPath(ast)

  const findImported: TransformHelper['findImported'] = (
    imp,
    loose = true,
    program = thisProgram
  ) => {
    // An import declaration can only be used in top-level.
    for (const path of program.get('body') as NodePath[]) {
      if (!path.isImportDeclaration()) continue
      if (matchImportStmt(imp, path.node, loose)) return path
    }
  }

  const hasImported: TransformHelper['hasImported'] = (
    imp,
    loose = true,
    program = thisProgram
  ) => {
    return findImported(imp, loose, program) !== undefined
  }

  // returns
  // - array: generated import statements
  // - path: all import statements are duplicated, returns the node path of the last one
  // - undefined: none import statement provided
  function normalizeImports(
    imports: ImportOption | ImportOption[],
    program: NodePath<Program>
  ) {
    if (Array.isArray(imports)) {
      if (imports.length === 0) return
    } else {
      imports = [imports]
    }
    // remove duplicated
    const toBeImported = imports.filter(
      (imp) => !hasImported(imp, false, program)
    )
    // all import statements are duplicated, returns the node path of the last one
    if (!toBeImported.length)
      return findImported(imports[imports.length - 1], false, program)!
    return template.statements.ast(
      Array.from(
        new Set(toBeImported.map((imp) => renderImportStmt(imp)))
      ).join('; ')
    ) as ImportDeclaration[]
  }

  const prependImports: TransformHelper['prependImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!toBeImported) return
    if (!Array.isArray(toBeImported)) return toBeImported
    const firstImport = (program.get('body') as NodePath[]).filter((p) =>
      p.isImportDeclaration()
    )[0]
    return (
      firstImport
        ? firstImport.insertBefore(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const appendImports: TransformHelper['appendImports'] = (
    imports,
    program = thisProgram
  ) => {
    const toBeImported = normalizeImports(imports, program)
    if (!toBeImported) return
    if (!Array.isArray(toBeImported)) return toBeImported
    const lastImport = (program.get('body') as NodePath[])
      .filter((p) => p.isImportDeclaration())
      .pop()
    return (
      lastImport
        ? lastImport.insertAfter(toBeImported)
        : program.unshiftContainer('body', toBeImported)
    ).pop() as NodePath<ImportDeclaration>
  }

  const prependToBody: TransformHelper['prependToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.unshiftContainer('body', nodes).pop() as NodePath
  }

  const appendToBody: TransformHelper['appendToBody'] = (
    nodes,
    program = thisProgram
  ) => {
    return program.pushContainer('body', nodes).pop() as NodePath
  }

  const getProgram: TransformHelper['getProgram'] = (node) => {
    if (!node) return thisProgram
    return node.findParent((p) => p.isProgram()) as NodePath<Program>
  }

  return {
    findImported,
    hasImported,
    prependImports,
    appendImports,
    prependToBody,
    appendToBody,
    getProgram,
  }
}
