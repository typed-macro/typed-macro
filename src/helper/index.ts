export * from './macro'
export * from './babel'
export type { ImportOption } from './import'
export type { DevServerHelper } from './server'
