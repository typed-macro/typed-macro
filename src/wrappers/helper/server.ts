import { ModuleNode, ViteDevServer } from 'vite'

export type DevServerHelper = {
  /**
   * Invalidate caches of modules that import macros from specific namespace,
   * only available in dev mode.
   *
   * This is usually used when, the macro needs to be re-expanded due to
   * the change of external conditions, while the modules that call
   * the macro cannot be automatically reloaded because their dependencies
   * in the module dependency graph have no change so Vite always caches them.
   */
  invalidateCache: (namespace: string) => void
  /**
   * Find modules that import macros from specific namespace,
   * only available in dev mode.
   */
  findImporters: (namespace: string) => Set<ModuleNode>
}

export function getDevServerHelper(server: ViteDevServer): DevServerHelper {
  function findImporters(ns: string) {
    // prefix `/@id/` is added by plugin `vite:import-analysis`
    // see:
    // https://github.com/vitejs/vite/blob/da0abc59935/packages/vite/src/node/plugins/importAnalysis.ts#L214
    return new Set(server.moduleGraph.getModuleById(`/@id/${ns}`)?.importers)
  }

  function invalidateCache(ns: string) {
    findImporters(ns)?.forEach((m) => server.moduleGraph.invalidateModule(m))
  }

  return {
    findImporters,
    invalidateCache,
  }
}
