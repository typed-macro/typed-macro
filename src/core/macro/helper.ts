import { ImportOption } from '@/core/helper/import'
import { Node, NodePath } from '@babel/traverse'
import { CallExpression, ImportDeclaration, Program } from '@babel/types'
import {
  normalizePathPattern,
  PathPatternNormalizer,
  projectDir,
} from '@/core/helper/path'
import {
  containsMacros,
  findImported,
  ImportedMacro,
} from '@/core/helper/traverse'
import {
  appendImports,
  appendToBody,
  prependImports,
  prependToBody,
} from '@/core/helper/transform'
import { promise } from '@/common'

export type MacroHelper = {
  /**
   * Prepend import statements to program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement,
   *   or undefined if imports param is array and length === 0
   */
  prependImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Append import statements to the last import statement of the program.
   * @param imports an import or an array of imports
   * @param program node path of the target program for prepending import statements.
   * use the one currently being handled by default.
   * @return node path of the last inserted import statement
   *   or undefined if imports param is array and length === 0
   */
  appendImports: (
    imports: ImportOption[] | ImportOption,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Find an import statement that has been in the target program already.
   * Note that findImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  findImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => NodePath<ImportDeclaration> | undefined

  /**
   * Check if an import statement has been in the target program already.
   * Note that hasImported() will not resolve extension for moduleName automatically.
   * @param imports an import
   * @param loose match loosely, defaults to true, e.g:
   * `import { a as _a } from 'a'` will match `{ moduleName: 'a' }` and `{ moduleName: 'a', exportName: 'a' }`
   * @param program node path of the target program. use the one currently being handled by default.
   */
  hasImported: (
    imports: ImportOption,
    loose?: boolean,
    program?: NodePath<Program>
  ) => boolean

  /**
   * Prepend any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  prependToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Append any node to the target program.
   * @param nodes any node or an array of nodes
   * @param program node path of the target program. use the one currently being handled by default.
   * @return node path of the last inserted node
   */
  appendToBody: <Nodes extends Node | readonly Node[]>(
    nodes: Nodes,
    program?: NodePath<Program>
  ) => NodePath

  /**
   * Normalize path pattern so can resolve file paths as import paths
   * @param path relative path or path pattern
   * @param root absolute file path of the project dir, default to projectDir('leaf')
   * @param importer absolute file path of the importer, default to current file
   *
   * e.g. search files by glob pattern and then import them
   *
   * ```typescript
   * // /src/example/a.ts
   * // /src/example/b.ts
   * // in /src/another/index.ts
   * const pattern = '../example/*.ts'
   * const { normalized, base, resolveImportPath } = normalizePathPattern(pattern)
   * const importPaths = glob.sync(normalized, { cwd: base }).map(resolveImportPath)
   *
   * // importPaths = ['../example/a.ts', '../example/b.ts']
   * ```
   */
  normalizePathPattern: (
    pattern: string,
    root?: string,
    importer?: string
  ) => PathPatternNormalizer

  /**
   * Get the directory of the root project or the nearest project.
   * @param which 'root' or 'leaf'
   */
  projectDir: (which: 'root' | 'leaf') => string

  /**
   * Check whether there are unexpanded macros under the path(s).
   * @param path the node path(s) to be checked, use the arguments of the current
   * call expression by default
   */
  containsMacros: (...paths: NodePath[]) => boolean[]

  /**
   * Yield current transformer to nested macros inside current
   * call expression and then call the current macro handler again
   * if the nested macros exist.
   * In another words, expand nested macros in arguments first.
   *
   * It is recommended to call this function at the beginning of the handler
   * that needs to ensure that the arguments don't contain nested macros,
   * to reduce unnecessary repeated calculations.
   *
   * e.g.
   * ```typescript
   * ({ path }, _, { yieldToNestedMacros }) => {
   *   yieldToNestedMacros()
   *   // ...your code
   * }
   * ```
   *
   * Internally, in v0.2.x, it throws a Promise, so DO NOT call it in a try block.
   */
  yieldToNestedMacros: () => never | void
}

export function createHelper(
  thisPath: NodePath<CallExpression>,
  thisProgram: NodePath<Program>,
  thisFilepath: string,
  importedMacros: ImportedMacro[]
) {
  const helpers: MacroHelper = {
    findImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose)
    },

    hasImported: (imp, loose = true, program = thisProgram) => {
      return findImported(program, imp, loose) !== undefined
    },

    prependImports: (imports, program = thisProgram) => {
      return prependImports(program, imports)
    },

    appendImports: (imports, program = thisProgram) => {
      return appendImports(program, imports)
    },

    prependToBody: (nodes, program = thisProgram) => {
      return prependToBody(program, nodes)
    },

    appendToBody: (nodes, program = thisProgram) => {
      return appendToBody(program, nodes)
    },

    normalizePathPattern: (
      pattern,
      root = projectDir('leaf'),
      importer = thisFilepath
    ) => {
      return normalizePathPattern(pattern, importer, root)
    },

    projectDir,

    containsMacros: (...paths) => {
      return containsMacros(
        paths.length ? paths : thisPath.get('arguments'),
        importedMacros
      )
    },

    yieldToNestedMacros: () => {
      if (
        containsMacros(thisPath.get('arguments'), importedMacros).includes(true)
      )
        throw promise
    },
  }

  return helpers
}
