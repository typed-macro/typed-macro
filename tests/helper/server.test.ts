import { getDevServerHelper } from '@/helper/server'
import { withDevServer } from '../testutils'

describe('DevServerHelper', () => {
  it('should work', async () => {
    await withDevServer(async (server) => {
      server.pluginContainer.resolveId = async (id) => ({ id })
      // see:
      // https://github.com/vitejs/vite/blob/da0abc59935/packages/vite/src/node/plugins/importAnalysis.ts#L214
      const source = await server.moduleGraph.ensureEntryFromUrl('/@id/source')
      const user = await server.moduleGraph.ensureEntryFromUrl('/@id/user')
      source.importers.add(user)
      user.importedModules.add(source)

      const helper = getDevServerHelper(server)
      expect(helper.findImporters('source')).toMatchSnapshot()
      expect(() => helper.invalidateCache('source')).not.toThrow()
    })
  })
})
