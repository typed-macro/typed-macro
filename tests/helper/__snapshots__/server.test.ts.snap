// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevServerHelper should work 1`] = `
Set {
  ModuleNode {
    "acceptedHmrDeps": Set {},
    "file": "/@id/user",
    "id": "/@id/user",
    "importedModules": Set {
      ModuleNode {
        "acceptedHmrDeps": Set {},
        "file": "/@id/source",
        "id": "/@id/source",
        "importedModules": Set {},
        "importers": Set {
          [Circular],
        },
        "isSelfAccepting": false,
        "lastHMRTimestamp": 0,
        "ssrModule": null,
        "ssrTransformResult": null,
        "transformResult": null,
        "type": "js",
        "url": "/@id/source",
      },
    },
    "importers": Set {},
    "isSelfAccepting": false,
    "lastHMRTimestamp": 0,
    "ssrModule": null,
    "ssrTransformResult": null,
    "transformResult": null,
    "type": "js",
    "url": "/@id/user",
  },
}
`;
